### PR TITLE
Fix recursion bomb in the k-induction algorithm

### PR DESCRIPTION
**Pull request by mikhailramalho**
The k-induction algorithm tries to get all the variables inside every loop, however, we were only checking for recursion where a function calls itself, not if we have a recursion cycle from function a -> b -> a.

This PR fixes the issue by keeping a list of function calls and checking if we already entered a given function before. We also pop the function call from the list when we hit a end-of-function.

Fixes issue #248.

### CODE REVIEW

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]
Three minor comments:

1) Could you please add a test case to this PR? e.g., a test case that succeeds and another one that fails
2) Could you also please rebase your branch? Otherwise, you won't be able to merge it into master.
3) Test case `digital-controller` is failing due to this PR.

**Comment by mikhailramalho** [comment]
New fix solves issues #248  and #252

**Comment by pmatos** [comment]
The fix works - `numarith.c` from racket seems to be verified without crashes with k-induction.

File: regression/esbmc/github_248/test.desc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+KNOWNBUG
```
Comment (lucasccordeiro) [inline regression/esbmc/github_248/test.desc:1]: Should it be declared as `CORE`?

**Review by lucasccordeiro — APPROVED** [review]
@mikhailramalho: I have just one minor comment for this PR.

**Pull request merged**